controllers/machinedeletionremediation_controller.go

```diff
@@ -282,23 +279,41 @@ func (r *MachineDeletionRemediationReconciler) getNodeFromMdr(ctx context.Contex
 	return node, nil
 }
 
-// getMachine retrieves a machine from the cluster based on the remediation.
-// It returns the machine, a boolean indicating whether the machine was expected to be found,
-// and an error if any occurred during the retrieval process.
+// getMachine retrieves a Machine from the cluster based on the remediation.
+// It returns the machine and an error if any occurred during the retrieval process.
 func (r *MachineDeletionRemediationReconciler) getMachine(ctx context.Context, remediation *v1alpha1.MachineDeletionRemediation) (*machinev1beta1.Machine, error) {
-	// If the Machine's Name and Ns come from the related Node, it is expected
-	// to find the Machine in the cluster, while if its Name and Ns come from
-	// CR's annotation, the Machine might have been deleted upon our request.
+	// If the remediation does not have the MachineNameNsAnnotation, we will try to get the Machine's data from one of
+	// the following sources:
+	// - the remediation's ownerReference, if the remediation was created by MachineHealthcheck
+	// - the remediation's Node, if the remediation was created by NodeHealthcheck
+
 	machineName, machineNs, err := getRemediationDataFromAnnotation(remediation, MachineNameNsAnnotation)
 	if err != nil {
 		r.Log.Error(err, "could not get Machine data from remediation", "remediation", remediation.GetName(), "annotation", MachineNameNsAnnotation)
 		return nil, unrecoverableError
 	}
 
 	var mustExist bool
+
+	if machineName == "" && len(remediation.GetOwnerReferences()) > 0 {
+		mustExist = true
+
+		machineKind, machineNamespace, err := r.getMachineKindAndNamespace(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, owner := range remediation.GetOwnerReferences() {
+			if owner.Kind == machineKind {
+				machineName, machineNs = owner.Name, machineNamespace
+				mustExist = true
+				r.Log.Info("MHC created CR: target machine is CR ownerReference", "machine", machineName, "namespace", machineNs)
+				break
+			}
+		}
+	}
+
 	if machineName == "" {
```

Comment:
How about `if-else` clauses here?
When `machineName == "" && len(remediation.GetOwnerReferences()) > 0` then you enter both of the _if_s.
Is it possible to enter `getMachine` when `machineName != ""`?

Reply:
>  Is it possible to enter getMachine when machineName != ""?

No, we only try to get `machineName` (and `machineNs`) if the previous attempt returned an empty string.

First we try to get this data from the CR's annotation.
If the annotation does not exist `machineName` is empty, so we try to get it from the `ownerReference` (if the `Kind` is "Machine"). Finally, if the `machineName` is still empty we try to get it from the Node name (`== CR's name`)

This means that we enter in [this if branch](https://github.com/medik8s/machine-deletion-remediation/blob/c1512270a5a51afb69d4dcbef54d5dc9e2492fd3/controllers/machinedeletionremediation_controller.go#L316) only if _both_ previous 2 attempts did not return a valid `machineName`

Note that returning an empty `machineName` is not an error by itself, so I cannot use the returned `error` to decide what to do.

Reply:
Got it.

>Note that returning an empty machineName is not an error by itself, so I cannot use the returned error to decide what to do.

You can concatenate the current error with a string of your choice for this issue, in case it helps :)

Reply:
Ah sure :), but what I meant is that returning an error would be incorrect